[WB-7810] Add Redis Module

elainaRenee opened the pull request:
Jira 

https://wandb.atlassian.net/browse/WB-7810

Adds a redis module so that we can externalize the filemeta queue. There are some other prs here that are prerequisites to making this possible

1. [Allow using redis connection string for filemeta queue](https://github.com/wandb/core/pull/8623)
2. [Add redis var to wandb terraform](https://github.com/wandb/terraform-kubernetes-wandb/pull/1)

I've tested the redis module and tested that the output can be passed to wandb local. I still need to test deploying my local branch in AWS using the external redis instance for the filemeta queue to make sure there are no connection/network issues.

elainaRenee commented on main.tf:
```diff
@@ -8,12 +8,12 @@ module "kms" {
-  kms_key_arn             = module.kms.key.arn
-  enable_external_storage = var.bucket_name != ""
+  kms_key_arn            = module.kms.key.arn
+  provision_file_storage = var.bucket_name == ""
```
thoughts on this name change? I was getting a bit confused with the `enable_external_storage` because local already has a concept of internal storage (minio) and external storage (s3,gcs,etc.)

jsbroks replied:
Way better! thank you!

vanpelt replied:
I'm still a little confused here.  When I see `file` storage I think persistent volume.  I gather what this flags is whether we are managing the bucket or the customer has brought their own.  It's probably fine, but my nits would be: `var.bucket_name => var.external_bucket_name` and `provision_file_storage => provision_bucket`

elainaRenee replied:
Sounds good, I'll update to `provision_bucket`. Are we ok to change variable names? I don't have a good sense of who is using this terraform and how versioning is set up (looks like we just reference the latest in deployer so worried it would be a breaking change)

jsbroks replied:
Yes it will be a break change. We'll need to manually migrate each customer. I have a PR up that will introduce proper versioning into this repo (currently we are using latest). Lets leave it out of this PR for now.

If we rename this variable then we should also reconsider naming all `file_storage` related resources, which once again are more breaking changes, and require more manual migrations. We should spend some time going through the tf code and renaming variables once versioning is properly configured.

vanpelt replied:
All good, I don't think added clarity justifies the pain of migrating things we can just keep these names.

elainaRenee commented on main.tf:
```diff
@@ -154,3 +154,9 @@ resource "aws_autoscaling_attachment" "autoscaling_attachment" {
 }
+
+module "redis" {
+  count  = var.use_internal_queue ? 1 : 0
```
I'm wondering if we need `use_internal_queue` to be a variable? If we get passed a bucket name, then we know not to provision file storage and use the redis queue. Unless there are instances where we would want to provision file storage and still use the redis queue?

jsbroks replied:
I though we are going to always provision the redis instance. It is used for other stuff then just the internal queue

vanpelt replied:
Let's just add a var that disables redis provisioning and is set to true by default.  We ideally want redis for caching and we'll need it for HA so it doesn't make sense to tie this directly to `use_internal_queue`

elainaRenee commented on modules/redis/main.tf:
```diff
@@ -0,0 +1,16 @@
+  port                          = 6379
+  maintenance_window            = var.preferred_maintenance_window
+  engine_version                = local.redis_version
+}
```
The [specs](https://docs.aws.amazon.com/AmazonElastiCache/latest/mem-ug/nodes-select-size.html) on the node type are ~ 6 GB of memory and 2 cores. I chose redis version 4 since that is what we use in prod currently. For failover it should automatically propagate the dns of the new primary node to the primary endpoint.

jsbroks replied:
Is version 6 an option? @annirudh @vanpelt I would perfer to use version 6 if we can

vanpelt replied:
Yes, let's use the latest version it should be wire compatible.  We should probably upgrade prod as 4 has [been EOL](https://docs.redis.com/latest/rs/administering/product-lifecycle/) for a while.

6 GB to start may be a little over kill, but is probably fine.  I would be curious to see the cost difference between a 3/6GB machine and this to make an informed decision.

vanpelt replied:
Also, is this magically added to our VPC and only accessible from it?  I didn't see any security group / network configuration.  Ideally this only listens on and allows connections from our VPC.

elainaRenee replied:
Good question, I am not sure about the networking piece yet. I plan to look into this today and test it (I'll post a link here to the instance as well). There is another node type `cache.t2.medium` that provisions an instance with 3 GB and is about half the cost (.068 per hour compared to .156 per hour)

vanpelt replied:
Cool, let's start with a cache.t2.medium and we can scale as needed.

jsbroks reviewed (changes requested):
I think we should always provision the redis instances, and not make it conditional if we are using an internal queue or not.

elainaRenee commented:
Addressed feedback, and tested in aws that my local instance (elaina.wandb.io) connects to the provisioned elasticache instance (I've been using the public-dns-external example). The pr right now assumes you're creating a vpc however - should I add the needed terraform / test the case when not creating a vpc?

vanpelt commented:
We should definitely support external redis in both private endpoint and internet modes, but I can't think of why we would ever be running the redis server outside of our VPC in either case.  As far as I knew we have two private flags currently.  One to make k8s only listen internally, and another to make the load balancer listen internally.  In both cases all of the infra should be in a VPC.

zythosec commented on modules/redis/main.tf:
```diff
@@ -0,0 +1,42 @@
+    protocol         = "tcp"
+    from_port        = "6379"
+    to_port          =  "6379"
+    cidr_blocks      = ["0.0.0.0/0"]
```
Should this have a fully open ingress security group? Or should this be limited to the VPC passed in? And then for the different examples, pass in the appropriate CIDR blocks? (same question for `ipv6_cidr_blocks`)

elainaRenee replied:
I pass the `vpc_id` here and updated to pass the relevant CIDR blocks (I also removed the `ipv6_cidr_blocks` option since we don't enable ipv6 when creating a vpc currently)

zythosec commented on modules/redis/main.tf:
```diff
@@ -0,0 +1,42 @@
+    ipv6_cidr_blocks = ["::/0"]
+  }
+
+  egress {
```
Same question as ingress. Does this really need to be fully open on all ports?

From terraform docs for aws_security_group:

Setting protocol = "all" or protocol = -1 with from_port and to_port will result in the EC2 API creating a security group rule with all ports open. This API behavior cannot be controlled by Terraform and may generate warnings in the future.

elainaRenee replied:
fixed

zythosec commented on variables.tf:
```diff
@@ -155,6 +155,12 @@ variable "network_database_subnet_cidrs" {
   default     = ["10.10.20.0/24", "10.10.21.0/24"]
 }
 
+variable "network_elasticache_subnet_cidrs" {
```
It doesn't look like this is used to actually restrict anything. Is this meant to be passed in to the redis submodule and used in the CIDR blocks?

elainaRenee replied:
yep it's passed into the networking module to create the subnet, which gets passed to the redis module

zythosec reviewed (changes requested):
@elainaRenee @jsbroks left some questions/comments regarding baseline security requirements.

elainaRenee commented on modules/redis/main.tf:
```diff
@@ -0,0 +1,42 @@
+  automatic_failover_enabled    = true
+  multi_az_enabled              = true
+  maintenance_window            = var.preferred_maintenance_window
+
```
nvm, made gorilla changes

elainaRenee commented:
Made some changes in gorilla to enable tls for our redis connection and built/pushed/deployed an image to aws and it works! 
![image](https://user-images.githubusercontent.com/22207683/153096071-b4f8eb7f-94ad-4401-ad4c-e2925d1ceb5b.png)
Any other concerns with this pr?

zythosec reviewed (approved):
Looks good to me! Can you add a ticket to the backlog to add auth_token to the aws_elasticache_replication_group using a rotating secret via secrets manager?

elainaRenee commented:
@vanpelt do you have any other comments here?

vanpelt reviewed (approved):
🔥

jsbroks commented:
This PR is included in version 1.0.0 :tada:
